### PR TITLE
Sort reported 'similar impl candidates' alphabetically

### LINKED ISSUE #53302: report_similar_impl_candidates seems to not put things in a canonical order
[I hit](https://github.com/rust-lang/rust/pull/52602) the following UI test failure in travis today: 
```diff
 	   |         ^^^^^^^^ the trait `std::convert::From<&str>` is not implemented for `i32`
 	   |
 	   = help: the following implementations were found:
+	             <i32 as std::convert::From<u8>>
 	             <i32 as std::convert::From<i8>>
-	             <i32 as std::convert::From<bool>>
 	             <i32 as std::convert::From<u16>>
 	             <i32 as std::convert::From<i16>>
-	             <i32 as std::convert::From<u8>>
+	             <i32 as std::convert::From<bool>>
 	   = note: required by `std::convert::From::from`
```
https://travis-ci.org/rust-lang/rust/jobs/415272826#L1942-L1958

Despite being fully-up-to-date against master and that passing locally.

Perhaps the order is deterministic normally, but different for me because I built `--incremental`?

I figured it was worth filing, at least, as I can't see any obvious sort in the relevant code:

https://github.com/rust-lang/rust/blob/5a0d2961ce88f7db90b13771d1e8fc3b50ded7b1/src/librustc/traits/error_reporting.rs#L407-L470

### PR DESCRIPTION
**Pull request by BurntPizza**
Fixes #53302

Thanks to @estebank for the mentor-like info.

### CODE REVIEW

File: src/test/ui/report-similar-candidates-ordering.rs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// issue-53302
```
Comment (estebank) [inline src/test/ui/report-similar-candidates-ordering.rs]: It seems like you added this comment after running `./x.py test src/test/ui --bless`:

```
[00:46:07] ---- [ui] ui/report-similar-candidates-ordering.rs stdout ----
[00:46:07] diff of stderr:
[00:46:07] 
[00:46:07] 1	error[E0277]: the trait bound `i32: std::convert::From<&str>` is not satisfied
[00:46:07] -	  --> $DIR/report-similar-candidates-ordering.rs:12:21
[00:46:07] +	  --> $DIR/report-similar-candidates-ordering.rs:14:21
[00:46:07] 3	   |
[00:46:07] 4	LL |     let _: i32 = "".into(); //~ ERROR [E0277]
[00:46:07] 5	   |                     ^^^^ the trait `std::convert::From<&str>` is not implemented for `i32`
```

**Review by estebank — APPROVED** [review]
Other than fixing the test, r=me.

Thanks for fixing this!

**Comment by BurntPizza** [comment]
ಠ_ಠ

**Comment by kennytm** [comment]
@bors r-

Merge conflict

**Comment by BurntPizza** [comment]
Rebased. Doing a local test run just to make sure I merged right.

**Comment by BurntPizza** [comment]
Oh yeah, a bunch of tests moved into ui, didn't they. 20 "failures."

**Comment by BurntPizza** [comment]
Turns out this was subsumed as a part of #53196. Closing.

**Pull request closed without merging**